### PR TITLE
[JUJU-871] Fix occasional race condition in worker/upgradedatabase/worker_test.go

### PR DESCRIPTION
**Pull request by jack-w-shaw**
Do this by increasing the timeout

Example of a failure is here: https://jenkins.juju.canonical.com/view/ci-run/job/Unit-RunUnitTests-race-amd64/1186/

Also remove a redundant EXPECT from the same test

Include a flyby bugfix as well, resolving a potential memory leak

## QA steps

Run unit tests in Jenkins multiple times to be confident the test now succeeds

## Documentation changes

N/A

## Bug reference

N/A

### CODE REVIEW

**Review by wallyworld — COMMENTED** [review]
Thanks for fixing this. Though I'm not sure it's the "right" fix. The test suite fully uses mocks so should be fast to run. `LongWait` should be more than enough time to wait for some mocked methods to be called.

**Comment by jack-w-shaw** [comment]
I'm not convinced either, but a race condition like this is the only thing I can see being a cause. Unless there's some other rare race condition I haven't spotted yet?

**Review by manadart — CHANGES_REQUESTED** [review]
If you look at this carefully you'll see that watching for primary changes introduced a new `clock.After` call that is implicitly satisfied by `s.clock.EXPECT().After(gomock.Any()).Return(timeout).AnyTimes()`.

This test needs needs tightening for expectations. I would.
- See to the parameter detection for the `After` expectation.
- Replace `AnyTimes` with `MinTimes(1)` where appropriate.
- Just close the `finished` channel instead of writing to it.

**Comment by jack-w-shaw** [comment]
Good spot Joe. I had noticed myself that the `EXPECT` line there catches both the main timeout, and the check primary loop, but hadn't put together quite how error prone that could be with both these timers pulling from the same channel.

Will fix now, and make the other clean ups you've pointed out

File: worker/upgradedatabase/worker_test.go [hunk 1]
```diff
@@ -256,25 +256,33 @@ func (s *workerSuite) TestNotPrimaryWatchForCompletionTimeout(c *gc.C) {
...
 	changes := make(chan struct{}, 2)
 	changes <- struct{}{}
 	changes <- struct{}{}
 	s.watcher.EXPECT().Changes().Return(changes).AnyTimes()
 
 	timeout := make(chan time.Time, 1)
-	s.clock.EXPECT().After(gomock.Any()).Return(timeout).AnyTimes()
+	s.clock.EXPECT().After(10 * time.Minute).Return(timeout)
+
+	neverTimeout := make(chan time.Time, 1)
```
Comment (manadart) [inline worker/upgradedatabase/worker_test.go:277]: Nit: don't need a buffer.

**Review by manadart — APPROVED** [review]
I think we'll find that this is the fix. Please update the description before landing.

**Comment by jack-w-shaw** [comment]
$$merge$$

**Comment by wallyworld** [comment]
merge job looks like it failed due to an issue in suite teardown, even though individual tests passed.
This will need fixing before landing.

**Comment by jack-w-shaw** [comment]
I suspect my flyby bug fix resolves this

**Comment by jack-w-shaw** [comment]
$$merge$$

**Pull request merged**